### ISSUE TITLE
Investigate jupyter-nb-clear-output pre-commit hook

### ISSUE THREAD

**Issue by OliviaLynn** [lincc-frameworks/python-project-template#127]
_(Haven't had a lot of time to look into this, but don't want my recent notebook PR to imply this is fixed & let us forget this.)_

I've run into an issue where the `jupyter-nb-clear-output` pre-commit hook would skip all notebooks (and pass as "skipped"), but not actually clear any output. Running the commands it lists directly in a terminal also fails for me.

Other reported problems include:

> Also, I'm curious if you've had a chance to confirm that the pre-commit hook that will clear output from jupyter notebooks will pass the test immediately after hydrating a project from the template? I was running into an issue early on when I had a demo notebook where the pre-commit hook would find some small thing that had changed, and it would result in a failed pre-commit hook test for the very first commit.

_Originally posted by @drewoldag in https://github.com/lincc-frameworks/python-project-template/issues/99#issuecomment-1466424656_

**Comment by drewoldag** [comment]
Oh, this is an interesting find. I _think_ that the thing that you've got quoted there was taken care of by this PR #120 (Specifically this change: https://github.com/lincc-frameworks/python-project-template/pull/120/files#diff-fa015e9f4b59b6a5bcdd6170eccd000649938892772163fa4298443a15d97dbe) 

But it's interesting that there are cases where the output isn't cleared because it's skipped. 

I wonder if this is a case where the pre-commit hook needs the `always_run: true` thing added to the hook configuration.

**Comment by drewoldag** [comment]
I took a quick look and tried out the approach of using the `always_run` flag like so: 
```
  - repo: local
    hooks:
      - id: jupyter-nb-clear-output
        name: jupyter-nb-clear-output
        description: Clear output from Jupyter notebooks.
        files: \.ipynb$
        stages: [commit]
        language: system
        entry: jupyter nbconvert --ClearOutputPreprocessor.enabled=True --inplace
        always_run: true
```

However, it doesn't work as expected for the same reason that running the entry command directly on the command line doesn't work, namely, if a file name isn't passed in, jupyter produces an error message and returns a non-zero exit code. 

For any hook, pre-commit will take all the files that are staged for commit, filter it down to just those that match the `files` pattern, and then pass each of those to the `entry` command. 

But if we specify `always_run`, and there aren't any files that match the pattern in the commit, then in this hook, jupyter produces an error, because no files are provided.

The error message mentions that multiple files can be passed in on the command line like so:
```
Multiple notebooks can be given at the command line in a couple of
            different ways:

            > jupyter nbconvert notebook*.ipynb
            > jupyter nbconvert notebook1.ipynb notebook2.ipynb
```

But I wasn't able to get a working example using various globbings. The following works, but I wasn't able to get a more general glob to function.

This worked:
```
>> jupyter nbconvert --ClearOutputPreprocessor.enabled=True --inplace ./docs/notebooks/*.ipynb
```

This did not:
```
>> jupyter nbconvert --ClearOutputPreprocessor.enabled=True --inplace **/*.ipynb
```

**Comment by drewoldag** [comment]
I was able to reproduce something like what was reported, but I'm not 100% sure it's the same thing.

Starting point:
I created a new project created from the template and ran the following:
`git checkout -b init; git add .; git commit -m 'test'; pre-commit install;`.

Repro steps:

1. Run the demo notebook in docs/notebooks so that there are output cells.
2. Make an edit to any other file. I added a new entry in .gitignore.
3. `git add .gitignore`
4. `git commit -m 'test'`

Output:
Git status shows that two files are modified.
> $> git status
> On branch init_branch
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git restore <file>..." to discard changes in working directory)
>         modified:   .gitignore
>         modified:   docs/notebooks/intro_notebook.ipynb
> 
> no changes added to commit (use "git add" and/or "git commit -a")
> 
> $> git add .gitignore 
> 
> $> git commit -m 'test'
> [WARNING] Unstaged files detected.
> [INFO] Stashing unstaged files to /home/drew/.cache/pre-commit/patch1680730943-2174.
> ...
> jupyter-nb-clear-output..............................(no files to check)Skipped
> ...

The end result is that the output wasn't cleared from the jupyter notebook, and the changes to .gitignore were committed.

**Comment by drewoldag** [comment]
In the same new project in the last comment, if I do not stage the jupyter notebook, and run pre-commit manually this is the output I see:

> $> git status
> On branch init_branch
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git restore <file>..." to discard changes in working directory)
>         modified:   docs/notebooks/intro_notebook.ipynb
> 
> no changes added to commit (use "git add" and/or "git commit -a")
> 
> $> pre-commit run jupyter-nb-clear-output
> [WARNING] Unstaged files detected.
> jupyter-nb-clear-output..............................(no files to check)Skipped

The end result is that the output is not cleared from the notebook. 

However if I run pre-commit with the `--all-files` flag, the output is cleared. Like so: 

> $> git status
> On branch init_branch
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git restore <file>..." to discard changes in working directory)
>         modified:   docs/notebooks/intro_notebook.ipynb
> 
> no changes added to commit (use "git add" and/or "git commit -a")
> 
> $> pre-commit run --all-files jupyter-nb-clear-output
> jupyter-nb-clear-output..................................................Failed
> - hook id: jupyter-nb-clear-output
> - files were modified by this hook
> 
> [NbConvertApp] Converting notebook docs/notebooks/intro_notebook.ipynb to notebook
> [NbConvertApp] Writing 3130 bytes to docs/notebooks/intro_notebook.ipynb

The end result is that the output is cleared from the notebooks.

**Comment by drewoldag** [comment]
Based on what I saw (and wrote up) in the last two comments, I think that aspect of the pre-commit hook is working as expected. 

However, it's possible there's another case that I've failed to notice that is causing this unexpected behavior. I've reached out to @OliviaLynn for any additional context that can be provided. If there's no additional context, I'll likely close this out as "cannot reproduce" and we can reopen the issue if we see it happening again.
